File: test/data/prefs/preferences_repository_codec_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/data/prefs/preferences_repository.dart';
import 'package:openvitals/domain/preferences/activity_recording_dashboard_layout.dart';
import 'package:openvitals/domain/preferences/activity_recording_preferences.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// Characterization tests for the two hand-rolled codecs inside
/// [PreferencesRepository] that had **no tests at all** — despite being ~90 lines
/// of bespoke string parsing standing between real users and their saved
/// settings.
///
/// They are written against the **public** API (set → reload from a fresh
/// repository → get), not the private helpers, so they pin the observable
/// behaviour and stay valid across any restructuring of the internals. That is
/// the point: they exist so the codecs can be moved out of this God object
/// without silently eating somebody's dashboard layout.
///
/// Where a test looks like it is asserting something trivial, it is usually
/// pinning a *sentinel*: this codec encodes "no value" as `0`, and getting that
/// round-trip wrong turns "off" into "0 metres" — which is a very different
/// setting.
Future<PreferencesRepository> _newRepo([
  Map<String, Object> initial = const {},
]) async {
  SharedPreferences.setMockInitialValues(initial);
  final prefs = await SharedPreferences.getInstance();
  return PreferencesRepository(prefs);
}

/// Writes through one repository, then reads through a *fresh* one — so the
/// assertion goes through the encoder AND the decoder, not an in-memory cache.
Future<T> _roundTrip<T>(
  void Function(PreferencesRepository repo) write,
  T Function(PreferencesRepository repo) read,
) async {
  final writer = await _newRepo();
  write(writer);
  final prefs = await SharedPreferences.getInstance();
  return read(PreferencesRepository(prefs));
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  group('activity recording dashboard layout codec', () {
    test('every template normalizes to largeTop — the others are storage-only',
        () async {
      // Pinning what the code ACTUALLY does, which is not what the type suggests.
      // `ActivityRecordingDashboardLayout.normalized()` resolves its grid through
      // a hardcoded top-level constant (`_recordingDashboardTemplate = largeTop`)
      // rather than through the instance's own `template` field, so a stored
      // twoByFour comes back as largeTop. The recording dashboard only ever draws
      // largeTop, so this is a deliberate narrowing rather than a bug -- but the
      // enum, the field and the codec all still carry the other two values, and
      // `withTemplate()` takes a template and silently ignores it (it has no
      // callers). Written down here so nobody "fixes" the round-trip below into a
      // feature that does not exist.
      for (final template in ActivityRecordingDashboardTemplate.values) {
        final result = await _roundTrip(
          (repo) => repo.setActivityRecordingDashboardLayout(
            'running',
            ActivityRecordingDashboardLayout(template: template),
          ),
          (repo) => repo.activityRecordingDashboardLayout('running'),
        );
        expect(
          result.template,
          ActivityRecordingDashboardTemplate.largeTop,
          reason: 'stored as ${template.name}',
        );
      }
    });

    test('round-trips the field order', () async {
      const fields = [
        ActivityRecordingDashboardField.speed,
        ActivityRecordingDashboardField.heartRate,
        ActivityRecordingDashboardField.distance,
      ];

      final result = await _roundTrip(
        (repo) => repo.setActivityRecordingDashboardLayout(
          'running',
          const ActivityRecordingDashboardLayout(
            template: ActivityRecordingDashboardTemplate.twoByFour,
            fields: fields,
          ),
        ),
        (repo) => repo.activityRecordingDashboardLayout('running'),
      );

      // Order is the whole point of the layout — it is what the user dragged.
      expect(result.fields.take(fields.length), fields);
    });

    test('round-trips per-field sizes', () async {
      final result = await _roundTrip(
        (repo) => repo.setActivityRecordingDashboardLayout(
          'running',
          ActivityRecordingDashboardLayout(
            template: ActivityRecordingDashboardTemplate.largeTop,
            fields: const [
              ActivityRecordingDashboardField.heartRate,
              ActivityRecordingDashboardField.speed,
            ],
            sizes: {
              ActivityRecordingDashboardField.heartRate:
                  ActivityRecordingDashboardItemSize(columnSpan: 2, rowSpan: 2),
            },
          ),
        ),
        (repo) => repo.activityRecordingDashboardLayout('running'),
      );

      final heartRate = result.items.firstWhere(
        (item) => item.field == ActivityRecordingDashboardField.heartRate,
      );
      expect(heartRate.size.columnSpan, 2);
      expect(heartRate.size.rowSpan, 2);
    });

    test('layouts are per activity type, not global', () async {
      final writer = await _newRepo();
      writer.setActivityRecordingDashboardLayout(
        'running',
        const ActivityRecordingDashboardLayout(
          fields: [ActivityRecordingDashboardField.heartRate],
        ),
      );
      writer.setActivityRecordingDashboardLayout(
        'cycling',
        const ActivityRecordingDashboardLayout(
          fields: [ActivityRecordingDashboardField.cadence],
        ),
      );

      final prefs = await SharedPreferences.getInstance();
      final repo = PreferencesRepository(prefs);
      expect(repo.activityRecordingDashboardLayout('running').fields.first,
          ActivityRecordingDashboardField.heartRate);
      expect(repo.activityRecordingDashboardLayout('cycling').fields.first,
          ActivityRecordingDashboardField.cadence);
    });

    test('an unknown activity type falls back to the default layout', () async {
      final repo = await _newRepo();
      expect(
        repo.activityRecordingDashboardLayout('never-configured').template,
        const ActivityRecordingDashboardLayout().template,
      );
    });

    test('a corrupt stored string degrades to the default, never throws',
        () async {
      // The decoder parses a bespoke separator format. Garbage in the store --
      // a downgrade, a half-written value, a future format -- must not take the
      // recording screen down with it.
      for (final corrupt in [
        '',
        'NOT_A_TEMPLATE',
        'NOT_A_TEMPLATE|HEART_RATE:1x1',
        '|||',
      ]) {
        final repo = await _newRepo({
          'activity_recording_dashboard_layout_running': corrupt,
        });
        expect(
          () => repo.activityRecordingDashboardLayout('running'),
          returnsNormally,
          reason: 'corrupt: "$corrupt"',
        );
      }
    });

    test('an unknown field in a stored layout is dropped, not fatal', () async {
      // Forward compatibility: a layout written by a NEWER build may name a
      // field this build has never heard of.
      final repo = await _newRepo({
        'activity_recording_dashboard_layout_running':
            'LARGE_TOP|HEART_RATE:1x1,WARP_DRIVE_RPM:1x1',
      });

      final layout = repo.activityRecordingDashboardLayout('running');
      expect(layout.fields, contains(ActivityRecordingDashboardField.heartRate));
    });
  });

  group('activity recording preferences — the null sentinels', () {
    // These four are stored as 0 meaning "off"/"null". Losing the distinction
    // turns "no route-gap limit" into "a route-gap limit of zero metres", which
    // would break the drawn line on every single fix.
    test('null route gap survives a round-trip as null, not zero', () async {
      final result = await _roundTrip(
        (repo) => repo.setActivityRecordingPreferences(
          const ActivityRecordingPreferences(routeGapMeters: null),
        ),
        (repo) => repo.activityRecordingPreferences(),
      );
      expect(result.routeGapMeters, isNull);
    });

    test('null distance interval survives as null, not zero', () async {
      final result = await _roundTrip(
        (repo) => repo.setActivityRecordingPreferences(
          const ActivityRecordingPreferences(
            recordingDistanceIntervalMeters: null,
          ),
        ),
        (repo) => repo.activityRecordingPreferences(),
      );
      expect(result.recordingDistanceIntervalMeters, isNull);
    });

    test('null voice intervals survive as null, not zero', () async {
      final result = await _roundTrip(
        (repo) => repo.setActivityRecordingPreferences(
          const ActivityRecordingPreferences(
            voiceAnnouncementTimeIntervalMinutes: null,
            voiceAnnouncementDistanceIntervalMeters: null,
          ),
        ),
        (repo) => repo.activityRecordingPreferences(),
      );
      expect(result.voiceAnnouncementTimeIntervalMinutes, isNull);
      expect(result.voiceAnnouncementDistanceIntervalMeters, isNull);
    });

    test('a real value round-trips as itself, and is not read as null',
        () async {
      final result = await _roundTrip(
        (repo) => repo.setActivityRecordingPreferences(
          const ActivityRecordingPreferences(
            routeGapMeters: 50,
            recordingDistanceIntervalMeters: 10,
          ).normalized(),
        ),
        (repo) => repo.activityRecordingPreferences(),
      );
      expect(result.routeGapMeters, isNotNull);
      expect(result.recordingDistanceIntervalMeters, isNotNull);
    });

    test('the booleans and the timeout round-trip', () async {
      final result = await _roundTrip(
        (repo) => repo.setActivityRecordingPreferences(
          const ActivityRecordingPreferences(
            autoIdleEnabled: false,
            keepScreenOnDuringRecording: false,
            voiceAnnouncementsEnabled: false,
            restTimerBellEnabled: false,
            autoIdleTimeoutSeconds: 45,
          ).normalized(),
        ),
        (repo) => repo.activityRecordingPreferences(),
      );

      expect(result.autoIdleEnabled, isFalse);
      expect(result.keepScreenOnDuringRecording, isFalse);
      expect(result.voiceAnnouncementsEnabled, isFalse);
      expect(result.restTimerBellEnabled, isFalse);
      expect(result.autoIdleTimeoutSeconds, 45);
    });
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/data/repository/PreferencesRepositoryCodecTest.kt
package tech.mmarca.openvitals.data.repository

import android.content.Context
import android.content.SharedPreferences
import io.mockk.every
import io.mockk.mockk
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import tech.mmarca.openvitals.devices.FakeSharedPreferences
import tech.mmarca.openvitals.domain.preferences.ActivityRecordingDashboardField
import tech.mmarca.openvitals.domain.preferences.ActivityRecordingDashboardItemSize
import tech.mmarca.openvitals.domain.preferences.ActivityRecordingDashboardLayout
import tech.mmarca.openvitals.domain.preferences.ActivityRecordingDashboardTemplate
import tech.mmarca.openvitals.domain.preferences.ActivityRecordingPreferences

/**
 * Port of the Flutter `preferences_repository_codec_test.dart` suite:
 * characterization tests for the two hand-rolled string codecs inside
 * [PreferencesRepository]. They are written against the public API
 * (set -> reload from a fresh repository -> get), so the assertion goes
 * through the encoder AND the decoder, not an in-memory cache.
 */
class PreferencesRepositoryCodecTest {

    private fun contextFor(prefs: SharedPreferences): Context = mockk {
        every {
            getSharedPreferences(PreferencesRepository.PREFS_FILE, Context.MODE_PRIVATE)
        } returns prefs
    }

    private fun newRepo(initial: Map<String, String> = emptyMap()): Pair<PreferencesRepository, FakeSharedPreferences> {
        val prefs = FakeSharedPreferences()
        val editor = prefs.edit()
        initial.forEach { (key, value) -> editor.putString(key, value) }
        editor.apply()
        return PreferencesRepository(contextFor(prefs)) to prefs
    }

    /**
     * Writes through one repository, then reads through a fresh one — the
     * analog of the Dart `_roundTrip` helper.
     */
    private fun <T> roundTrip(
        write: (PreferencesRepository) -> Unit,
        read: (PreferencesRepository) -> T,
    ): T {
        val (writer, prefs) = newRepo()
        write(writer)
        return read(PreferencesRepository(contextFor(prefs)))
    }

    // region activity recording dashboard layout codec

    @Test fun `every template normalizes to LARGE_TOP - the others are storage-only`() {
        // Pinning what the code ACTUALLY does: `normalized()` resolves the grid
        // through the hardcoded LARGE_TOP recording template rather than the
        // instance's own `template` field, so a stored TWO_BY_FOUR comes back
        // as LARGE_TOP. Written down here so nobody "fixes" the round-trip into
        // a feature that does not exist.
        ActivityRecordingDashboardTemplate.entries.forEach { template ->
            val result = roundTrip(
                write = { repo ->
                    repo.setActivityRecordingDashboardLayout(
                        "running",
                        ActivityRecordingDashboardLayout(template = template),
                    )
                },
                read = { repo -> repo.activityRecordingDashboardLayout("running") },
            )
            assertEquals(
                "stored as ${template.name}",
                ActivityRecordingDashboardTemplate.LARGE_TOP,
                result.template,
            )
        }
    }

    @Test fun `round-trips the field order`() {
        val fields = listOf(
            ActivityRecordingDashboardField.SPEED,
            ActivityRecordingDashboardField.HEART_RATE,
            ActivityRecordingDashboardField.DISTANCE,
        )

        val result = roundTrip(
            write = { repo ->
                repo.setActivityRecordingDashboardLayout(
                    "running",
                    ActivityRecordingDashboardLayout(
                        template = ActivityRecordingDashboardTemplate.TWO_BY_FOUR,
                        fields = fields,
                    ),
                )
            },
            read = { repo -> repo.activityRecordingDashboardLayout("running") },
        )

        // Order is the whole point of the layout — it is what the user dragged.
        assertEquals(fields, result.fields.take(fields.size))
    }

    @Test fun `round-trips per-field sizes`() {
        val result = roundTrip(
            write = { repo ->
                repo.setActivityRecordingDashboardLayout(
                    "running",
                    ActivityRecordingDashboardLayout(
                        template = ActivityRecordingDashboardTemplate.LARGE_TOP,
                        fields = listOf(
                            ActivityRecordingDashboardField.HEART_RATE,
                            ActivityRecordingDashboardField.SPEED,
                        ),
                        sizes = mapOf(
                            ActivityRecordingDashboardField.HEART_RATE to
                                ActivityRecordingDashboardItemSize(columnSpan = 2, rowSpan = 2),
                        ),
                    ),
                )
            },
            read = { repo -> repo.activityRecordingDashboardLayout("running") },
        )

        val heartRate = result.items.first {
            it.field == ActivityRecordingDashboardField.HEART_RATE
        }
        assertEquals(2, heartRate.size.columnSpan)
        assertEquals(2, heartRate.size.rowSpan)
    }

    @Test fun `layouts are per activity type, not global`() {
        val (writer, prefs) = newRepo()
        writer.setActivityRecordingDashboardLayout(
            "running",
            ActivityRecordingDashboardLayout(
                fields = listOf(ActivityRecordingDashboardField.HEART_RATE),
            ),
        )
        writer.setActivityRecordingDashboardLayout(
            "cycling",
            ActivityRecordingDashboardLayout(
                fields = listOf(ActivityRecordingDashboardField.CADENCE),
            ),
        )

        val repo = PreferencesRepository(contextFor(prefs))
        assertEquals(
            ActivityRecordingDashboardField.HEART_RATE,
            repo.activityRecordingDashboardLayout("running").fields.first(),
        )
        assertEquals(
            ActivityRecordingDashboardField.CADENCE,
            repo.activityRecordingDashboardLayout("cycling").fields.first(),
        )
    }

    @Test fun `an unknown activity type falls back to the default layout`() {
        val (repo, _) = newRepo()
        assertEquals(
            ActivityRecordingDashboardLayout().template,
            repo.activityRecordingDashboardLayout("never-configured").template,
        )
    }

    @Test fun `a corrupt stored string degrades to the default, never throws`() {
        // The decoder parses a bespoke separator format. Garbage in the store —
        // a downgrade, a half-written value, a future format — must not take
        // the recording screen down with it.
        listOf(
            "",
            "NOT_A_TEMPLATE",
            "NOT_A_TEMPLATE|HEART_RATE=1x1",
            "|||",
        ).forEach { corrupt ->
            val (repo, _) = newRepo(
                mapOf("activity_recording_dashboard_layout_running" to corrupt),
            )
            val layout = runCatching { repo.activityRecordingDashboardLayout("running") }
            assertTrue("corrupt: \"$corrupt\"", layout.isSuccess)
            assertNotNull("corrupt: \"$corrupt\"", layout.getOrNull())
        }
    }

    @Test fun `an unknown field in a stored layout is dropped, not fatal`() {
        // Forward compatibility: a layout written by a NEWER build may name a
        // field this build has never heard of.
        val (repo, _) = newRepo(
            mapOf(
                "activity_recording_dashboard_layout_running" to
                    "LARGE_TOP|HEART_RATE=1x1,WARP_DRIVE_RPM=1x1",
            ),
        )

        val layout = repo.activityRecordingDashboardLayout("running")
        assertTrue(layout.fields.contains(ActivityRecordingDashboardField.HEART_RATE))
    }

    // endregion

    // region activity recording preferences — the null sentinels

    // These are stored as 0 meaning "off"/"null". Losing the distinction turns
    // "no route-gap limit" into "a route-gap limit of zero metres", which would
    // break the drawn line on every single fix.

    @Test fun `null route gap survives a round-trip as null, not zero`() {
        val result = roundTrip(
            write = { repo ->
                repo.setActivityRecordingPreferences(
                    ActivityRecordingPreferences(routeGapMeters = null),
                )
            },
            read = { repo -> repo.activityRecordingPreferences() },
        )
        assertNull(result.routeGapMeters)
    }

    @Test fun `null distance interval survives as null, not zero`() {
        val result = roundTrip(
            write = { repo ->
                repo.setActivityRecordingPreferences(
                    ActivityRecordingPreferences(recordingDistanceIntervalMeters = null),
                )
            },
            read = { repo -> repo.activityRecordingPreferences() },
        )
        assertNull(result.recordingDistanceIntervalMeters)
    }

    @Test fun `null voice intervals survive as null, not zero`() {
        val result = roundTrip(
            write = { repo ->
                repo.setActivityRecordingPreferences(
                    ActivityRecordingPreferences(
                        voiceAnnouncementTimeIntervalMinutes = null,
                        voiceAnnouncementDistanceIntervalMeters = null,
                    ),
                )
            },
            read = { repo -> repo.activityRecordingPreferences() },
        )
        assertNull(result.voiceAnnouncementTimeIntervalMinutes)
        assertNull(result.voiceAnnouncementDistanceIntervalMeters)
    }

    @Test fun `a real value round-trips as itself, and is not read as null`() {
        val result = roundTrip(
            write = { repo ->
                repo.setActivityRecordingPreferences(
                    ActivityRecordingPreferences(
                        routeGapMeters = 50,
                        recordingDistanceIntervalMeters = 10,
                    ).normalized(),
                )
            },
            read = { repo -> repo.activityRecordingPreferences() },
        )
        assertNotNull(result.routeGapMeters)
        assertNotNull(result.recordingDistanceIntervalMeters)
    }

    @Test fun `the booleans and the timeout round-trip`() {
        val result = roundTrip(
            write = { repo ->
                repo.setActivityRecordingPreferences(
                    ActivityRecordingPreferences(
                        autoIdleEnabled = false,
                        keepScreenOnDuringRecording = false,
                        voiceAnnouncementsEnabled = false,
                        restTimerBellEnabled = false,
                        autoIdleTimeoutSeconds = 45,
                    ).normalized(),
                )
            },
            read = { repo -> repo.activityRecordingPreferences() },
        )

        assertFalse(result.autoIdleEnabled)
        assertFalse(result.keepScreenOnDuringRecording)
        assertFalse(result.voiceAnnouncementsEnabled)
        assertFalse(result.restTimerBellEnabled)
        assertEquals(45, result.autoIdleTimeoutSeconds)
    }

    // endregion
}
